File: java-penguin/src/main/java/com/function/Action.java
package com.function;

import java.util.HashMap;
import java.util.Map;

public class Action {

    private static final  String ROTATE_RIGHT = "rotate-right";
    private static final  String ROTATE_LEFT = "rotate-left";
    private static final  String ADVANCE = "advance";
    private static final  String RETREAT = "retreat";
    private static final  String SHOOT = "shoot";
    private static final  String PASS = "pass";
    private static final  Map<String, String> MOVE_UP;
    private static final  Map<String, String> MOVE_DOWN;
    private static final  Map<String, String> MOVE_RIGHT;
    private static final  Map<String, String> MOVE_LEFT;

    static {
        MOVE_UP = new HashMap<>();
        MOVE_UP.put("top", ADVANCE);
        MOVE_UP.put("bottom", ROTATE_LEFT);
        MOVE_UP.put("left", ROTATE_RIGHT);
        MOVE_UP.put("right", ROTATE_LEFT);

        MOVE_DOWN = new HashMap<>();
        MOVE_DOWN.put("top", ROTATE_LEFT);
        MOVE_DOWN.put("bottom", ADVANCE);
        MOVE_DOWN.put("left", ROTATE_LEFT);
        MOVE_DOWN.put("right", ROTATE_RIGHT);

        MOVE_RIGHT = new HashMap<>();
        MOVE_RIGHT.put("top", ROTATE_RIGHT);
        MOVE_RIGHT.put("bottom", ROTATE_LEFT);
        MOVE_RIGHT.put("left", ROTATE_LEFT);
        MOVE_RIGHT.put("right", ADVANCE);

        MOVE_LEFT = new HashMap<>();
        MOVE_LEFT.put("top", ROTATE_LEFT);
        MOVE_LEFT.put("bottom", ROTATE_RIGHT);
        MOVE_LEFT.put("left", ADVANCE);
        MOVE_LEFT.put("right", ROTATE_RIGHT);
    }

    private Match match;

    private String moveTowardsCenterOfMap()
    {
        int centerPointx = match.mapWidth / 2;
        int centerPointy = match.mapHeight / 2;
        return moveTowardsPoint(centerPointx, centerPointy);
    }

    private String moveTowardsPoint(int pointx, int pointy)
    {
        int  penguinPositionx = match.you.x;
        int penguinPositiony = match.you.y;
        String plannedAction = PASS;

        if (penguinPositionx < pointx)
        {
            plannedAction = MOVE_RIGHT.get(match.you.direction);
        }
        else if (penguinPositionx > pointx)
        {
            plannedAction = MOVE_LEFT.get(match.you.direction);
        }
        else if (penguinPositiony < pointy)
        {
            plannedAction = MOVE_DOWN.get(match.you.direction);
        }
        else if (penguinPositiony > pointy)
        {
            plannedAction = MOVE_UP.get(match.you.direction);
        }
        if (plannedAction.equals(ADVANCE) && wallInFrontOfPenguin())
        {
            return SHOOT;
        }
        return plannedAction;
    }

    private boolean doesCellContainWall(int x, int y)
    {
        for (int i = 0; i < match.walls.length; i++) {
            if (match.walls[i].x == x && match.walls[i].y == y) {
                return true;
            }
        }
        return false;
    }

    private boolean wallInFrontOfPenguin()
    {
        switch (match.you.direction)
        {
            case "top":
                return doesCellContainWall(match.you.x, --match.you.y);
            case "bottom":
                return doesCellContainWall(match.you.x, ++match.you.y);
            case "left":
                return doesCellContainWall(--match.you.x, match.you.y);
            case "right":
                return doesCellContainWall(++match.you.x, match.you.y);
            default:
                return true;
        }
    }

    Action(Match match) {
        this.match = match;
    }

    String chooseAction() {
        String response;
        response = moveTowardsCenterOfMap();
        return response;
    }
}

File: csharp-penguin/Shared/Action.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CSharpPenguin.Shared
{
    public class Action
    {
        readonly string ROTATE_RIGHT = "rotate-right";
        readonly string ROTATE_LEFT = "rotate-left";
        readonly string ADVANCE = "advance";
        readonly string RETREAT = "retreat";
        readonly string SHOOT = "shoot";
        readonly string PASS = "pass";
        readonly Dictionary<string, string> MOVE_UP = new Dictionary<string, string> { { "top", "advance" }, { "bottom", "rotate-left" }, { "right", "rotate-left" }, { "left", "rotate-right" } };
        readonly Dictionary<string, string> MOVE_DOWN = new Dictionary<string, string> { { "top", "rotate-left" }, { "bottom", "advance" }, { "right", "rotate-right" }, { "left", "rotate-left" } };
        readonly Dictionary<string, string> MOVE_RIGHT = new Dictionary<string, string> { { "top", "rotate-right" }, { "bottom", "rotate-left" }, { "right", "advance" }, { "left", "rotate-right" } };
        readonly Dictionary<string, string> MOVE_LEFT = new Dictionary<string, string> { { "top", "rotate-left" }, { "bottom", "rotate-right" }, { "right", "rotate-left" }, { "left", "advance" } };
        private readonly Match _match;

        public Action(Match match)
        {
            _match = match;
        }
        public string ChooseAction()
        {
            var response = PASS;
            response = MoveTowardsCenterOfMap();
            return response;
        }

        string MoveTowardsCenterOfMap()
        {
            var centerPointX = (int)_match.MapWidth / 2;
            var centerPointY = (int)_match.MapHeight / 2;
            return MoveTowardsPoint(centerPointX, centerPointY);
        }

        string MoveTowardsPoint(int pointX, int pointY)
        {
            var penguinPositionX = _match.You.X;
            var penguinPositionY = _match.You.Y;
            var plannedAction = PASS;

            if (penguinPositionX < pointX)
            {
                plannedAction = MOVE_RIGHT[_match.You.Direction];
            }
            else if (penguinPositionX > pointX)
            {
                plannedAction = MOVE_LEFT[_match.You.Direction];
            }
            else if (penguinPositionY < pointY)
            {
                plannedAction = MOVE_DOWN[_match.You.Direction];
            }
            else if (penguinPositionY > pointY)
            {
                plannedAction = MOVE_UP[_match.You.Direction];
            }
            if (plannedAction == ADVANCE && WallInFrontOfPenguin())
            {
                return SHOOT;
            }
            return plannedAction;
        }

        bool DoesCellContainWall(int x, int y)
        {
            return _match.Walls.Where(wall => wall.X == x && wall.Y == y).ToList().Count > 0;
        }

        bool WallInFrontOfPenguin()
        {
            switch (_match.You.Direction)
            {
                case "top":
                    return DoesCellContainWall(_match.You.X, --_match.You.Y);
                case "bottom":
                    return DoesCellContainWall(_match.You.X, ++_match.You.Y);
                case "left":
                    return DoesCellContainWall(--_match.You.X, _match.You.Y);
                case "right":
                    return DoesCellContainWall(++_match.You.X, _match.You.Y);
                default:
                    return true;
            }
        }
    }
}
